File: python/test/simple.py
import os
import sys

ovf_py_dir = os.path.abspath(os.path.join(os.path.dirname( __file__ ), ".."))
sys.path.insert(0, ovf_py_dir)

from ovf import ovf

import numpy as np

# import unittest

##########

print("----- ovf test nonexistent")
with ovf.ovf_file("nonexistent.ovf") as ovf_file:
    print("found:      ", ovf_file.found)
    print("is_ovf:     ", ovf_file.is_ovf)
    print("n_segments: ", ovf_file.n_segments)
    segment = ovf.ovf_segment()
    if ovf_file.read_segment_header(0, segment) != -1:
        print("read_segment_header failed: ", ovf_file.get_latest_message())
print("----- ovf test nonexistent done")


print("----- ovf test writing")
with ovf.ovf_file(os.path.abspath(os.path.join(os.path.dirname( __file__ ), "testfile_out.ovf"))) as ovf_file:
    data = np.zeros((2, 2, 1, 3), dtype='d')
    data[0,1,0,:] = [3.0, 2.0, 1.0]
    segment = ovf.ovf_segment(n_cells=[2,2,1])
    if ovf_file.write_segment(segment, data) != -1:
        print("write_segment failed: ", ovf_file.get_latest_message())
    data[0,1,0,:] = [4.0, 5.0, 6.0]
    if ovf_file.append_segment(segment, data) != -1:
        print("append_segment failed: ", ovf_file.get_latest_message())
print("----- ovf test writing done")


print("----- ovf test reading")
with ovf.ovf_file(os.path.abspath(os.path.join(os.path.dirname( __file__ ), "testfile_out.ovf"))) as ovf_file:
    print("found:      ", ovf_file.found)
    print("is_ovf:     ", ovf_file.is_ovf)
    print("n_segments: ", ovf_file.n_segments)
    segment = ovf.ovf_segment()
    if ovf_file.read_segment_header(0, segment) != -1:
        print("read_segment_header failed: ", ovf_file.get_latest_message())
    data_shape = (segment.n_cells[0], segment.n_cells[1], segment.n_cells[2], 3)
    print("data shape: ", data_shape)
    data = np.zeros(data_shape, dtype='f')
    if ovf_file.read_segment_data(0, segment, data) != -1:
        print("read_segment_data failed: ", ovf_file.get_latest_message())
    print("first segment:  ", data[0,1,0,:])
    if ovf_file.read_segment_data(1, segment, data) != -1:
        print("read_segment_data failed: ", ovf_file.get_latest_message())
    print("second segment: ", data[0,1,0,:])
print("----- ovf test reading done")

File: fortran/test/simple.f90
program main
use ovf

    type(ovf_file)      :: file
    type(ovf_segment)   :: segment
    integer             :: success
    real(kind=4), allocatable :: array_4(:,:)
    real(kind=8), allocatable :: array_8(:,:)

    call file%open_file("python/test/testfile_py.ovf")
    write (*,*) "Found      = ", file%found
    write (*,*) "is_ovf     = ", file%is_ovf
    write (*,*) "n_segments = ", file%n_segments

    success = file%read_segment_header(segment)
    if ( success == OVF_OK) then
        write (*,*) "n_cells = ", segment%N_Cells
        write (*,*) "n_total = ", segment%N
    else
        write (*,*) "test read_segment_header did not work. Message: ", file%latest_message
        STOP 1
    endif

    ! Allocation of arrays is optional with ovf read functions
    ! allocate( array_4(3, segment%N) )
    ! allocate( array_8(3, segment%N_Cells) )

    success = file%read_segment_data(segment, array_4)
    if ( success == OVF_OK) then
        write (*,*) "array_4(:,2) = ", array_4(:,2)
    else
        write (*,*) "test read_segment_data on array_4 did not work. Message: ", file%latest_message
        STOP 1
    endif

    success = file%read_segment_data(segment, array_8)
    if ( success == OVF_OK) then
        write (*,*) "array_8(:,2) = ", array_8(:,2)
    else
        write (*,*) "test read_segment_data on array_8 did not work. Message: ", file%latest_message
        STOP 1
    endif

    success = file%close_file()
    if ( success == OVF_OK) then
        write (*,*) "test file closed"
    else
        write (*,*) "test close_file did not work. Message: ", file%latest_message
        STOP 1
    endif

end program main